### ISSUE TITLE
HTTP/2 connection pool can create extra connections unnecessarily

### ISSUE THREAD

**Issue by trustin** [line/armeria#818]
When an HTTP/2 connection is acquired from a `KeyedChannelPool`, it is removed from the pool:

- https://github.com/line/armeria/blob/armeria-0.54.1/core/src/main/java/com/linecorp/armeria/client/pool/DefaultKeyedChannelPool.java#L128

and then added back after the initial request headers are written:

- https://github.com/line/armeria/blob/armeria-0.54.1/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L168

When another client makes a request at the same time, a new HTTP/2 connection may be created unnecessarily, because the connection acquired for the first request does not exist in the pool.

Instead of removing the connection completely from the pool's data structure, we could count the number of active requests for each connection and return the one with the least number of active requests, which is similar to what `EventLoopScheduler` does.

For even more efficiency, we could merge `EventLoopScheduler` and `DefaultKeyedChannelPool` into one and defer the selection of `EventLoop` as late as possible. Then, we won't need to count per-`EventLoop` active requests at all too choose the most idle `EventLoop`; we can just find the connection with the least active requests and that will get us the right `EventLoop`.

Related discussion: https://github.com/line/armeria/issues/816

**Comment by anuraaga** [comment]
I haven't ever used `ClientRequestContext.eventLoop` ever. Since my client is always used from a server handler, I would always schedule a client's callbacks onto the server event loops. Not worrying about breaking the API SGTM.

By the way, did the event loops scheduler change remove the propagation of server event loop to client context? I can't find where the server loop is propagated to the client anymore.

**Comment by trustin** [comment]
@anuraaga It's in `HttpClientFactory.eventLoopSupplier`.

**Comment by anuraaga** [comment]
Right but I couldn't find it being used by the client. Since client (`UserClient` or `ArmeriaChannel`) use `acquireEventLoop`, don't they ignore `eventLoopSupplier` / `RequestContext.current()` entirely?

**Comment by trustin** [comment]
@anuraaga Indeed. We lost that optimization when we introduced `EventLoopScheduler`. Let me try to get it back while resolving this issue.

**Comment by trustin** [comment]
Related: #800

**Comment by imasahiro** [comment]
The breaking API sounds good to me too. 
(And actually, I faced extra connection when I write a batch that calls http2+thrift (and it causes "Too many open files" error..). So it would be happy reducing connection as much as possible)

**Comment by anuraaga** [comment]
Thinking about this more, have we made it too complicated? I really liked the simplicity and ease of reasoning of this approach we used to use to select the client's event loop.

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java#L96

I think this is quite optimal for determining client connections from a server context, everything stays in the same thread. Is my understanding correct that the `EventLoopScheduler` is mainly useful for when a client is used from a non-armeria server context, e.g. legacy server or batch job? Can we just limit use of the scheduler to when there is no `RequestContext`? Keeping things simple and optimal for the armeria server case is probably our priority.

**Comment by trustin** [comment]
Now I remember why we do not have that optimization anymore. `eventLoopSupplier` was actually causing the connection pool to create as many connections as `numServerEventLoops * numEndpoints`, which can grow really fast when `numServerEventLoops` is large. Hence, it also defeats the advantage of multiplex connections. We could have a 'preference' based on the current `EventLoop`, though.
